package/src/bin/omicron-package.rs

```diff
@@ -926,6 +926,62 @@ impl Config {
             _ => bail!("Aborting"),
         }
     }
+
+    /// Returns target packages to be assembled on the builder machine, limited
+    /// to those specified in `only` (if set).
+    // TODO: should this be moved to omicron-zone-package?
+    fn packages_to_build(&self) -> PackageMap<'_> {
+        let packages = self.package_config.packages_to_build(&self.target);
+        if self.only.is_empty() {
+            packages
+        } else {
+            // create a new PackageMap and copy over the selected packages
+            let mut filtered_packages = PackageMap(BTreeMap::new());
+            for package_name in &self.only {
+                if let Some(package) = packages.0.get(package_name) {
+                    filtered_packages.0.insert(package_name, package);
+                }
+            }
+            // recursively find and copy over the dependencies
+            let lookup_by_output = packages
+                .0
+                .iter()
+                .map(|(name, package)| {
+                    (package.get_output_file(name), (*name, *package))
+                })
+                .collect::<BTreeMap<_, _>>();
+            loop {
+                let to_add = filtered_packages
+                    .0
+                    .values()
+                    .filter_map(|package| {
+                        if let PackageSource::Composite { packages } =
+                            &package.source
+                        {
+                            Some(packages)
+                        } else {
+                            None
+                        }
+                    })
+                    .flatten()
+                    .map(|output| {
+                        *lookup_by_output.get(output).unwrap_or_else(|| {
+                            panic!(
+                                "Could not find a package which creates '{}'",
+                                output
+                            )
+                        })
+                    })
+                    .collect::<Vec<_>>();
```

Comment:
I think this is correct, but on each subsequent iteration, we don't need to look at *all* `filtered_pacakges` again, right? only the ones we just added?

(Like, this needs to do a walk of all the leaf nodes in the dependency tree, but it is actually re-walking the whole tree on each iteration I think)